src/iam/certhandler/modules/pkcs11.cpp

```diff
@@ -0,0 +1,986 @@
+        return GeneratePIN(cLoginTypeGroup, cTeeClientUUIDNs, gid, userPIN);
+    }
+
+    LOG_ERR() << "Wrong TEE login type=" << loginType;
+    return AOS_ERROR_WRAP(Error::Enum::eInvalidArgument);
+}
+
+Error PKCS11Module::GeneratePIN(const String& loginType, const uuid::UUID& space, uint32_t data, String& userPIN)
+{
+    StaticString<pkcs11::cPINLength> dataBuf;
+    uuid::UUID                       sha1;
+    StaticString<cUUIDStringLen>     sha1Str;
+
+    auto err = dataBuf.Format("uid: %" PRIx64, data);
+    if (!err.IsNone()) {
+        AOS_ERROR_WRAP(err);
+    }
+
+    Tie(sha1, err)
+        = mUUIDManager->CreateSHA1(space, Array<uint8_t>(reinterpret_cast<uint8_t*>(dataBuf.Get()), dataBuf.Size()));
+    if (!err.IsNone()) {
+        AOS_ERROR_WRAP(err);
+    }
+
+    Tie(sha1Str, err) = mUUIDManager->UUIDToString(sha1);
+    if (!err.IsNone()) {
+        AOS_ERROR_WRAP(err);
+    }
+
+    return userPIN.Format("%s:%s", loginType.CStr(), sha1Str.CStr());
+    ;
+};
+
+Error PKCS11Module::GetUserPin(String& pin) const
+{
+    if (!mTeeLoginType.IsEmpty()) {
+        pin.Clear();
+        return Error::Enum::eNone;
+    }
+
+    return FS::ReadFile(mConfig.mUserPINPath, pin);
+}
+
+RetWithError<UniquePtr<pkcs11::SessionContext>> PKCS11Module::CreateSession(bool userLogin, const String& pin)
+{
+    using pkcs11::SessionInfo;
+
+    Error                             err = Error::Enum::eNone;
+    UniquePtr<pkcs11::SessionContext> session;
+
+    Tie(session, err) = mPKCS11->OpenSession(mSlotID, CKF_RW_SESSION | CKF_SERIAL_SESSION);
+    if (!err.IsNone()) {
+        return {nullptr, err};
+    }
+
+    LOG_DBG() << "Create session. session=" << session->GetHandle() << ", slotID=" << mSlotID;
+
+    SessionInfo sessionInfo;
+
+    err = session->GetSessionInfo(sessionInfo);
+    if (!err.IsNone()) {
+        return {nullptr, err};
+    }
+
+    const bool isUserLoggedIn
+        = sessionInfo.state == CKS_RO_USER_FUNCTIONS || sessionInfo.state == CKS_RW_USER_FUNCTIONS;
+    const bool isSOLoggedIn = sessionInfo.state == CKS_RW_SO_FUNCTIONS;
+
+    if ((userLogin && isSOLoggedIn) || (!userLogin && isUserLoggedIn)) {
+        err = session->Logout();
+        if (!err.IsNone()) {
+            return {nullptr, err};
+        }
+    }
+
+    if (userLogin && !isUserLoggedIn) {
+        LOG_DBG() << "User login. session=" << session->GetHandle() << ", slotID=" << mSlotID;
+
+        return {Move(session), session->Login(CKU_USER, mUserPIN)};
+    }
+
+    if (!userLogin && !isSOLoggedIn) {
+        LOG_DBG() << "SO login. session=" << session->GetHandle() << ", slotID=" << mSlotID;
+
+        return {Move(session), session->Login(CKU_SO, pin)};
+    }
+
+    return {Move(session), Error::Enum::eNone};
+}
+
+Error PKCS11Module::FindObject(pkcs11::SessionContext& session, const SearchObject& filter, Array<SearchObject>& dst)
+{
+    static constexpr auto cSearchObjAttrCount = 4U;
+
+    CK_BBOOL token = CK_TRUE;
```

Comment:
`auto`?

Reply:
I would keep the type explicit in case if left / right sides are of different types.